projects/packages/waf/src/class-waf-standalone-bootstrap.php

```diff
@@ -0,0 +1,144 @@
+<?php
+/**
+ * Handles generation and deletion of the bootstrap for the standalone WAF mode.
+ *
+ * @package automattic/jetpack-waf
+ */
+
+namespace Automattic\Jetpack\Waf;
+
+use Composer\InstalledVersions;
+use Exception;
+
+/**
+ * Handles the bootstrap.
+ */
+class WafStandaloneBootstrap {
+
+	/**
+	 * Ensures that constants are initialized if this class is used.
+	 */
+	public function __construct() {
+		$this->guard_against_missing_abspath();
+		$this->initialize_constants();
+	}
+
+	/**
+	 * Ensures that this class is not used unless we are in the right context.
+	 *
+	 * @return void
+	 * @throws Exception If we are outside of WordPress.
+	 */
+	private function guard_against_missing_abspath() {
+
+		if ( ! defined( 'ABSPATH' ) ) {
+			throw new Exception( 'Cannot generate the WAF bootstrap if we are not running in WordPress context.' );
+		}
+	}
+
+	/**
+	 * Initializes the constants required for generating the bootstrap, if they have not been initialized yet.
+	 *
+	 * @return void
+	 */
+	private function initialize_constants() {
+		if ( ! defined( 'JETPACK_WAF_DIR' ) ) {
+			define( 'JETPACK_WAF_DIR', trailingslashit( WP_CONTENT_DIR ) . 'jetpack-waf' );
+		}
+	}
+
+	/**
+	 * Initialized the WP filesystem and serves as a mocking hook for tests.
+	 *
+	 * @return void
+	 */
+	protected function initialize_filesystem() {
```

Comment:
Is WP_Filesystem functionality that unit tests can't correctly implement?

Reply:
@ArSn answered this - our unit tests don't set up an instance of WordPress so obviously, no, they can't implement WP_Filesystem. This question came about because in Hiring, the Code Test unit tests automatically spin up a fresh WordPress instance to run tests against.